bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/AsyncTableWrapper.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.mirroring.hbase1_x;
+
+import com.google.api.core.InternalApi;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import java.util.List;
+import java.util.concurrent.Callable;
+import org.apache.hadoop.hbase.client.Get;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.Table;
+
+/**
+ * MirroringClient verifies consistency between two databases asynchronously - after the results are
+ * delivered to the user. HBase Table object does not have an synchronous API, so we simulate it by
+ * wrapping the regular Table into AsyncTableWrapper.
+ *
+ * <p>Table instances are not thread-safe, every operation is synchronized to prevent concurrent
+ * accesses to the table from different threads in the executor.
+ */
+@InternalApi("For internal usage only")
+public class AsyncTableWrapper {
+  private final Table table;
+  private final ListeningExecutorService executorService;
+
+  public AsyncTableWrapper(Table table, ListeningExecutorService executorService) {
+    this.table = table;
+    this.executorService = executorService;
+  }
+
+  public ListenableFuture<Result> get(final Get gets) {
+    return this.executorService.submit(
+        new Callable<Result>() {
+          @Override
+          public Result call() throws Exception {
+            synchronized (table) {
+              return table.get(gets);
```

Comment:
The `table` object is not thread safe, and we issue a single asynchronous operation that is synchronized on this `table` object per single blocking client operation. If the secondary database has the same performance as the primary, then we think that is shouldn't really matter, because this synchronized operation will more-or-less overlap with next synchronous operation issued by the user thread. If it is not the case we will hit flow controller limit anyways.